DESIGN_GUIDELINES.md

```diff
@@ -1165,6 +1226,87 @@ use the SecretValue type [_awslint:secret-token_].
+the bucket.
+
+Finally, in rare cases, the AWS Construct Library includes some even
+higher-level constructs, which we call patterns. These constructs are
+designed to help you complete common tasks in AWS, often involving multiple
+kinds of resources. For example, the
+`aws-ecs-patterns.ApplicationLoadBalancedFargateService` construct represents an
+architecture that includes an AWS Fargate container cluster employing an
+Application Load Balancer (ALB). These examples are rare, as these patterns
+constructs are, by design, heavily opinionated and less customizable; this means
+while the patterns are useful for many customers as-is, many others will find
+the patterns not _quite_ right and inflexible to their needs.
+
+#### What Belongs in the CDK?
+
+What divides the L2s from the higher-level constructs, and where do we draw the
+line on what belongs as part of the CDK itself?
+
+Generally speaking, an L2 directly models a single CloudFormation resource. The
+L2 typically exposes all of the underlying surface area of the underlying L1, as
+well as provides convenience methods for interactions with other services, or
+creating related or dependent resources, permissions, and metrics.
+
+Examples of behaviors that an L2 commonly include:
+
+- Modeling of the underlying L1 properties
+- Methods for adding related resources (e.g., adding an event notification to
+  an S3 bucket).
+- Modeling of permissions and resource policies
+- Modeling of metrics
+
+In addition to the above, some higher-order L2s may introduce more complex and
+helpful functionality, either part of the original L2 itself, or as part of a
+separate construct. The most common form of these L2s are integration constructs
+that model interactions between different services (e.g., SNS publishing to SQS,
+CodePipeline actions that trigger Lambda functions).
+
+A higher-level L2 may also be included in the CDK if another L2 requires or
+would greatly benefit from a dependency on it.
+
+As general guidelines, the following types of constructs likely don't belong in
+the CDK, and should be published as separate artifacts:
+
+- Constructs that create complete application infrastructure spanning multiple
+  resources (e.g., networking + compute). These constructs are often very
+  helpful to a specific subset of customers, but require high levels of
+  customization to work for all customers.
+- Constructs that extend the base L2s to provide functionality for a specific
+  language (e.g., `PythonFunction`), operating system (e.g., `WindowsInstance`),
+  or framework (e.g., `FlaskRestApi`). Rationale: subclassing the L2s in the CDK
+  leads to awkward inheritence chains, especially for customers who themselves
+  subclass the constructs to enforce organizational standards (e.g., `Bucket` ->
+  `SecureBucket`). Prefer to integrate this support directly into the
+  base L2 -- either directly or via composition -- where possible.
```

Comment:
I think the fact that customers subclass in order to enforce policies is actually the anti-pattern (albeit I am aware it's what people are doing), but I think we should address that at the framework level (see https://github.com/aws/aws-cdk-rfcs/issues/25) and not sacrifice our APIs because of that.

I think `WindowsInstance` and `PythonFunction` fall under the L1.5 category and do belong in the library. Trying to model all these APIs into the base L2 sometimes results in very poor APIs which are basically the "union" of all the sub-types, and this is exactly the case for L1.5s

Reply:
I totally agree that aws/aws-cdk-rfcs#25 is the way to solve policies, but that's not yet an option for customers, and I thought it was worth acknowledging the current state in these guidelines.

> Trying to model all these APIs into the base L2 sometimes results in very poor APIs which are basically the "union" of all the sub-types, and this is exactly the case for L1.5s

Counter-argument. Isn't the behavior of `PythonFunction` (or `NodeJsFunction`) what we'd like the default behavior of `Function` to be? Why *wouldn't* we want to offer this functionality as part of the base experience? Looking at the props and implementation of `PythonFunction` (for example), it's not clear why we couldn't have achieved the same via `lambda.Code.fromPythonSource(...)`.

Using inheritance over composition here runs the same danger as faced by L3s/patterns when we suddenly need more than one attribute. For example, we now have `Function`, `PythonFunction`, `NodeJsFunction`, `SingletonFunction`, and `EdgeFunction`. What if I want a singleton Edge function (#12498), singleton NodeJS function (#7564), or Python Edge function (#12671)? This becomes more complicated when we've chosen to subclass to provide functionality.

Reply:
The API of `PythonFunction` and `NodeJsFunction` is tailored to the respective language. If we try to put all that logic in `Function` we end up with mutually exclusive options and lose strong type checking.

Granted, the inheritance model makes it slightly difficult to compose but TBH the singleton function example is not very compelling. I'd argue that singletons are trivial to define (and functions are not the only singletons you'd want in a CDK app), and mostly should be regarded as a pattern and not as an abstraction. I'd be willing to deprecate it altogether in 2.x.

Reply:
Additionally, I don't think this discussion is about inheritance versus composition because both are legitimate for 1.5 and each case should be evaluated on its own.

The question is how do we define these 1.5s and whether they belong in the CDK.

Reply:
> Additionally, I don't think this discussion is about inheritance versus composition because both are legitimate for 1.5 and each case should be evaluated on its own.

I guess I _mostly_ agree that it's a tangent. However, if -- for example -- LambdaFunction and PythonFunction weren't implemented as their own L2(.5)s, but instead as extensions of the existing `Code` class, then they're not really L2.5s anymore. I guess maybe `Code` is an L2.5 at that point? But I'm unsure, because I think of L2.Xs as logically modeling/extending L1s, not helper classes.

That being said, I don't want to lose focus, so I'll disagree & commit on the Functions and simplify the language on L2.5s.